astropy/table/table.py

```diff
@@ -2262,6 +2262,21 @@ def __len__(self):
             self._first_colname = next(iter(self.columns))
             return len(self.columns[self._first_colname])
 
+    def __or__(self, other):
+        try:
+            updated_table = self.copy()
+            updated_table.update(other, copy=True)
+            return updated_table
+        except TypeError:
```

Comment:
You should test these exceptions.

Reply:
Sorry (first time doing this so I want to be super clear), do you mean I should update the tests to run code we expect to cause `TypeError`s and check that it returns `NotImplemented`?

Reply:
Yup!

Reply:
Just playing with this gives what might be a surprising result to some users. The result for `1 | t` makes sense, but then `t | 1` gives a `ValueError` related to not being able to create a Table with `1`.

In retrospect having the `update()` method auto-create a Table from the input seems like not the best idea, but we're stuck with the existing API for that method. 

BUT, for the new `|` and `|=` operators, maybe we should insist on `other` being a `Table` subclass and raise `TypeError` if not.

```
In [14]: 1 | t
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [14], in <module>
----> 1 1 | t

TypeError: unsupported operand type(s) for |: 'int' and 'Table'

In [15]: t | 1
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [15], in <module>
----> 1 t | 1

File ~/git/astropy/astropy/table/table.py:2268, in Table.__or__(self, other)
   2266 try:
   2267     updated_table = self.copy()
-> 2268     updated_table.update(other, copy=True)
   2269     return updated_table
   2270 except TypeError:

File ~/git/astropy/astropy/table/table.py:3413, in Table.update(self, other, copy)
   3410 from .operations import _merge_table_meta
   3412 if not isinstance(other, Table):
-> 3413     other = self.__class__(other, copy=copy)
   3414 common_cols = set(self.colnames).intersection(other.colnames)
   3415 for name, col in other.items():

File ~/git/astropy/astropy/table/table.py:852, in Table.__init__(self, data, masked, names, dtype, meta, copy, rows, copy_indices, units, descriptions, **kwargs)
    849     data = [[]] * n_cols
    851 else:
--> 852     raise ValueError(f"Data type {type(data)} not allowed to init Table")
    854 # Set up defaults if names and/or dtype are not specified.
    855 # A value of None means the actual value will be inferred
    856 # within the appropriate initialization routine, either from
    857 # existing specification or auto-generated.
    859 if dtype is None:

ValueError: Data type <class 'int'> not allowed to init Table
```

Reply:
I agree with testing for `Table` here, i.e., something like
```
if not isinstance(other, Table):
    return NotImplemented
...
```
With that, any errors should just be raised, so no need for `try/except`.